### PR TITLE
Fix global eval

### PR DESCRIPTION
**Pull request by CPFLAME**
这个pr要做的:

- [x] 对齐qqp的超参, 并跑出近似精度(libai 68.99 vs megatron 68.94)
- [x] 修复数据并行时, test中sampler的采样缺失问题, 以及test_batch_size为1时错误运行问题
- [x] 修复test阶段模型返回的是loss_dict问题
- [x] 修改sampler的unittest适配新sampler的逻辑
- [x] 解决evaluator中最后一个batch导致计算精度有偏差的问题
- [x] 将所有的consistent换成global
- [x] 跑通所有的单测

### 当前的evaluator输出
会输出最后计算指标时的有效样本总数(4315, 和len(dataset)一致)
已更换过多个test batch size进行测试(1, 2, 4, 8), 总样本数均为4315
image

### CODE REVIEW

File: libai/data/samplers/samplers.py [hunk 1]
```diff
@@ -169,11 +169,15 @@ def __iter__(self):
                 yield batch
                 batch = []
 
-        if len(batch) > 0 and not self.drop_last:
-            # TODO: fix sampler bug when batch_size=1
-            for _ in range(self.micro_batch_size - len(batch)):
+        if not self.drop_last:
+            if self.data_parallel_rank >= remain and remain > 0:
```
Comment (CPFLAME) [inline libai/data/samplers/samplers.py:173]: 在此处进行sampler补0的操作

File: libai/evaluation/cls_evaluator.py [hunk 2]
```diff
@@ -56,9 +56,8 @@ def reset(self):
         self._predictions = []
 
     def process(self, inputs, outputs):
-        # FIX ME: support dict args, not implement in graph right now
-        pred_logits = outputs[-1]  # decide by your model output
-        labels = inputs[-1]  # decide by your dataloder output
+        pred_logits = outputs["prediction_scores"]
+        labels = inputs["label"]
```
Comment (CPFLAME) [inline libai/evaluation/cls_evaluator.py:60]: 这里指定了默认值, 其他任务想调用cls_evaluator. 必须是的dataloader中存在key值`label` 以及 网络输出key值`prediction_scores`

File: libai/evaluation/evaluator.py [hunk 3]
```diff
@@ -148,6 +148,8 @@ def inference_on_dataset(
     total_compute_time = 0
     total_eval_time = 0
     consumed_samples = 0
+    dps = dist.get_data_parallel_size()
+    last_batch_lack = (dps - (total % dps)) % dps
```
Comment (CPFLAME) [inline libai/evaluation/evaluator.py:152]: 需要在此处进行判断最后的batch是否有缺失. 

不采用tensor batch size进行判断的原因是:
由于sampler中补0的情况, 可能使得tensor batch size恰好等于指定的batch, 从而误认为last batch的数据全部都为有效数据

File: libai/evaluation/evaluator.py [hunk 4]
```diff
@@ -209,6 +211,7 @@ def inference_on_dataset(
     total_time = time.perf_counter() - start_time
     total_time_str = str(datetime.timedelta(seconds=total_time))
     # NOTE this format is parsed by grep
+    logger.info("Total valid samples: {}".format(consumed_samples))
```
Comment (CPFLAME) [inline libai/evaluation/evaluator.py:214]: 添加输出有效样本数目, 方便自己验证

File: libai/evaluation/utils.py [hunk 5]
```diff
@@ -18,25 +18,35 @@
...
-        padded_list.append(padded_xi)
-    return padded_list, valid_sample
+        padded_xi[:tensor_batch, ...] = padded_xi[:tensor_batch, ...] + xi
+        padded_xi = padded_xi.to_consistent(
+            sbp=dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.broadcast])
+        )
+        for i in range(last_batch_lack - 1):
+            start_idx = tensor_micro_batch_size * (data_parallel_size - i - 1) - 1
+            padded_xi[start_idx:-1] = padded_xi[start_idx + 1 :]
+        padded_xi = padded_xi.to_consistent(sbp=xi.sbp)
```
Comment (CPFLAME) [inline libai/evaluation/utils.py:47]: 这里先把tensor变成broadcast的原因是, 
需要把有效的样本全部集中在tensor的头部, 
必须从tensor的尾部开始一个一个往前挪

File: libai/models/bert_model.py [hunk 6]
```diff
@@ -343,7 +343,7 @@ def forward(
         seq_relationship_score = self.seq_relationship(pooled_output)
         prediction_scores = self.lm_logits(prediction_scores, word_embeddings_weight)
 
-        if lm_labels is not None:
+        if self.training and lm_labels is not None:
```
Comment (CPFLAME) [inline libai/models/bert_model.py:346]: 由于当前graph的输入方式为**kwargs, 导致lm_labels为一直未非空, 所以要通过self.training来判断是否输出loss

**Review by rentainhe — APPROVED** [review]

File: libai/models/utils/graph_base.py [hunk 7]
```diff
@@ -59,6 +59,8 @@ def __init__(
                     # stage 3
                     flow.boxing.nccl.disable_group_boxing_by_dst_parallel(True)
             self.set_pipeline_stage_id()
+        else:
+            self.model.training = False
```
Comment (L1aoXingyu) [inline libai/models/utils/graph_base.py:63]: 应该会在别的地方调用 `model.eval()` ，需要在这里设置这个内容吗

File: projects/QQP/configs/config_qqp.py [hunk 8]
```diff
@@ -77,10 +84,10 @@
 
 graph = dict(
     enabled=True,
-    train_graph=LazyCall(ClassificationGraph)(
+    train_graph=LazyCall(GraphBase)(
         is_train=True,
         recompute_grad=True,
         fp16=True,
```
Comment (L1aoXingyu) [inline projects/QQP/configs/config_qqp.py:90]: `recompute_grad` 和 `fp16` 都会在 `train.py` 里面配置，然后会在代码里面处理，这样参数会在一个地方统一设置

File: libai/models/utils/graph_base.py (hunk 7 above)
Comment (CPFLAME) [reply libai/models/utils/graph_base.py:63]: 需要, 我没加的时候,
发现training仍然是`True`
所以在这里加了

Comment (L1aoXingyu) [reply libai/models/utils/graph_base.py:63]: 是不是应该在 evaluator inference 之前，调用 `model.eval()`

Comment (CPFLAME) [reply libai/models/utils/graph_base.py:63]: model是graph的, 直接用的`flow.no_grad()`,   graph也可以设置model.eval()吗?

Comment (CPFLAME) [reply libai/models/utils/graph_base.py:63]: 我先自己本地实验一下吧.

File: libai/evaluation/evaluator.py [hunk 9]
```diff
@@ -223,6 +229,12 @@ def inference_on_dataset(
         )
     )
 
+    # set model to train
+    if isinstance(model, flow.nn.Graph):
+        model.model.train()
+    elif isinstance(model, flow.nn.Module):
+        model.train()
```
Comment (L1aoXingyu) [inline libai/evaluation/evaluator.py:236]: 这里是不是应该根据之前模型的 training_mode 来设置，而不是统一设成 train，参考这里
https://github.com/facebookresearch/detectron2/blob/ef2c3abbd36d4093a604f874243037691f634c2f/detectron2/evaluation/evaluator.py#L221

**Review by L1aoXingyu — APPROVED** [review]
